CM-235: Support for istio-csr through integration with cert-manager-operator (#1628)
openshift/enhancements: enhancements/cert-manager/istio-csr-controller.md

  when required by creating the custom resource.
- As an OpenShift user, I want to have an option to dynamically configure `istio-csr`, so that only the required
  features can be enabled by updating the custom resource.
- As an OpenShift user, I should be able to disable `istio-csr` when not required by removing the custom resource,
  and controller should cleanup all resources created for the `istio-csr` deployment.
- As an OpenShift user, I should be able to install `istio-csr` in the upgrade cluster where istiod control plane
  is active and should be able to update the certificate endpoint to `istio-csr` endpoint.
- As an OpenShift user, I want to have an option to dynamically enable monitoring for the `istio-csr` project and
  to use the OpenShift monitoring solution when required.

### Goals

- `cert-manager-operator` to be extended to manage `istio-csr` along with currently managed `cert-manager`.
- New custom resource `istiocsr` to be made available to enable, configure and disable deployment.

### Non-Goals

- `istio-csr` can be used only with supported version of `OpenShift Service Mesh`. Please refer `Version Skew Strategy`
  section for more details.

## Proposal

> `istio-csr` will be managed by `cert-manager-operator` which can be enabled and configured by a user as and
> when required, using new custom resource which will be made available for configuring `istio-csr`.

Reviewer: ```suggestion
`istiocsr` will be installed and managed by `cert-manager-operator`. A new custom resource is defined to configure the `istiocsr`. The `istio-csr` CR can be added day2 to install `istiocsr` post the installation or upgrade of cert-manager operator on the existing OpenShift clusters.
```

A new controller will be added to `cert-manager-operator` to manage and maintain the `istio-csr` deployment in
desired state which can be monitored by user through the status sub-resource of the new custom resource.

`istio-csr` controller will make use of static manifest templates for creating below resources required for
successfully deploying `istio-csr` agent. Please refer `Implementation Details/Notes/Constraints` section for more details.
* `certificate.cert-manager.io`
* `clusterissuer.cert-manager.io`
* `issuer.cert-manager.io`
* `clusterrole.rbac.authorization.k8s.io`
* `clusterrolebinding.rbac.authorization.k8s.io`
* `deployment`
* `service`
* `servicemonitor.monitoring.coreos.com`
* `namespace`
* `role.rbac.authorization.k8s.io`
* `rolebinding.rbac.authorization.k8s.io`
* `serviceaccount`

> `istio-csr` CR object will be a singleton object, must be named `default` and an OpenShift user will have to
> create the same for enabling the deployment. `istio-csr` will be deployed in the namespace where the CR object is created.

Reviewer: ```suggestion
`istio-csr` CR object is a singleton object and is enforced in CRD to have name as `default`. `istiocsr` will be deployed in the namespace where the CR object is created.
```

> When `istiocsr` custom resource is created without defining `issuerref`, the operator will create a `ClusterIssuer` in the

Reviewer: As per https://github.com/openshift/enhancements/pull/1628/files#r1846779286, the condition of optional issuerref does not occur.

> When `istiocsr` custom resource is created without defining `issuerref`, the operator will create a `ClusterIssuer` in the

Reviewer: The usage of terms `istiocsr` and `istio-csr` is not consistent across the proposal. Also, a section that defines these terms will help to alleviate the ambiguity.

> `cert-manager` namespace using a self-signed `Issuer`, which will act as a root CA. And the `ClusterIssuer` will be used to

Reviewer: I do not think, I understand "And the `ClusterIssuer` will be used to create an `Issuer` an intermediate CA,"  this section needs clarification. A ClusterIssuer cannot be created with an Issuer or vice versa.

create an `Issuer` an intermediate CA, in the istio-csr deployed namespace and will used to obtain certificates for gRPC
service made available for `OpenShift Service Mesh` and also for issuing the certificates requested by `OpenShift Service
Mesh`. The `Issuer` in the istio-csr deployed namespace will be used for obtaining the signed certificates requested by 
the `OpenShift Service Mesh`.

> Configurations made available in the spec of `istiocsr` custom resource are passed as command line arguments to `istio-csr`

Reviewer: The configuration updates will make the system enter into an inconsistent state, for example, updates to certificate specfic configurations like issuer, trust bundles (CA configurations), algorithm updates. The configuration updates should only be allowed to low-impact fields. Can you please add more on how the operator intends to validate this flow?

agent and updating these configurations would cause new rollout of the `istio-csr` deployment, which means a new pod will
be created and old pod will terminated resulting in `istio-csr` restart.

> When an OpenShift user deletes `istiocsr` CR object `istio-csr` agent deployment will be deleted and all the

Reviewer: Adding auto-deletion, an un-intentional delete of CR will render istio and workloads when it coincides with certificate expiry or revocation. For the TP version, it can be considered to have user perform the un-installation by explicitly deleting respective resources. A label or annotation would be needed that eases query of the resources by user.

Author: This will be considered for GA.

other resources created will also be cleaned up, including the `clusterissuer` and `issuer` objects if created by the
controller.

A fork of [upstream](https://github.com/cert-manager/istio-csr) `istio-csr` will be created [downstream](https://github.com/openshift/cert-manager-istio-csr)
for better version management.

### Workflow Description

> - Enable istio-csr deployment

Reviewer: ```suggestion
- Installation of istio-csr
```

>   - An OpenShift user enables `istio-csr` by creating the new custom resource made available for configuring

Reviewer: "Enables" signifies that istio-csr agent is already installed. I believe, "installs" would be the right term.

  `istio-csr`.
  - `istio-csr` controller based on the configuration in CR
    - Creates below `Issuers` when user does not provide `issuerref` to be used by `istio-csr` for obtaining the certificates.
      - Creates an `ClusterIssuer` in `cert-manager` namespace which acts as root CA for obtaining the
        CA certificate for `istio-csr` use case.
      - Creates an `Issuer` in `istio-system` namespace which is an intermediate CA signed by `ClusterIssuer`
        root CA.
    - Deploys istio-csr agent in `istio-system` namespace.

![alt text](./istio-csr-create.png).

> - Disable istio-csr deployment

Reviewer: This section might need an update.

---
title: istio-csr-controller
authors:
  - "@bhb"
reviewers:
  - "@tgeer" ## reviewer for cert-manager component
approvers:
  - "@tgeer" ## approver for cert-manager component
api-approvers:
  - "@tgeer" ## approver for cert-manager component
creation-date: 2024-01-22
last-updated: 2024-05-13
tracking-link:
  - https://issues.redhat.com/browse/CM-234
see-also:
  - NA
replaces:
  - NA
superseded-by:
  - NA
---

# Extend cert-manager-operator to manage istio-csr.

## Summary

This enhancement describes the proposal to extend cert-manager-operator to deploy and manage the `istio-csr`

> solution. `OpenShift Service Mesh` facilitates the security of both intra-cluster and external service

Reviewer: An elaboration about the istio-csr is required. Would `istio-csr solution` mean a controller or set of controllers?

Author: A controller will be used manage the `istio-csr`.

communications through mTLS. And istio-csr acts as an agent to obtain and renew certificates that
facilitate mTLS from the `cert-manager`.

## Motivation

cert-manager enables obtaining certificates from [various](https://docs.openshift.com/container-platform/4.15/security/cert_manager_operator/index.html#cert-manager-issuer-types_cert-manager-operator-about)
certificate authorities and also ensures the validity of certificates by renewing them before they expire.
`OpenShift Service Mesh` secures intra-cluster and external service connections with TLS encryption. `cert-manager`
can be used to obtain the required certificates through custom resources `issuer` and `clusterissuer`. `istio-csr`
is an agent that simplifies the process of obtaining the certificates from the `cert-manager` for
`OpenShift Service Mesh`. `cert-manager-operator` manages `cert-manager` and extending the operator to manage
`istio-csr` will help the users to use all the solutions mentioned above effectively and easily.

### User Stories

> - As an OpenShift user, I want to have an option to dynamically enable `istio-csr`, so that it can be used only

Reviewer: A user story is required for installation, un-installation and upgrade usecases.

Author: comment incorporated.

- As an OpenShift user, I want to have an option to dynamically configure `istio-csr`, so that only the required
  features can be enabled.

### Goals

- `cert-manager-operator` to be extended to manage `istio-csr` along with currently managed `cert-manager`.
- New custom resource `istiocsr` to be made available to enable and configure deployment.

### Non-Goals

- `istio-csr` must be enabled and configured before installing `OpenShift Service Mesh`, which is a
  limitation of `OpenShift Service Mesh`.
- Downgrading OpenShift cluster from `istio-csr` supported version to unsupported version is not supported.

## Proposal

`istio-csr` will be managed by `cert-manager-operator` which can be enabled and configured by a user as and
when required, using new custom resource which will be made available for configuring `istio-csr`.

A new controller will be added to `cert-manager-operator` to manage and maintain the `istio-csr` deployment in
desired state which can be monitored by user through the status sub-resource of the new custom resource.

`istio-csr` controller will make use of static manifest templates for creating below resources to
successfully deploy `istio-csr` agent.

> * `certificate.cert-manager.io`

Reviewer: The list has resources created by cert-manager operator as part for cert-manager installation. Are there any differences that needs to be considered for istio-csr?

Author: These are specific to istio-csr, overlapping resources could be merged, but thought to be keep different for istio-csr, so that when istio-csr is disabled just these resources will be cleared.

* `clusterissuer.cert-manager.io`
* `issuer.cert-manager.io`
* `clusterrole.rbac.authorization.k8s.io`
* `clusterrolebinding.rbac.authorization.k8s.io`
* `deployment`
* `service`
* `servicemonitor.monitoring.coreos.com`
* `namespace`
* `role.rbac.authorization.k8s.io`
* `rolebinding.rbac.authorization.k8s.io`
* `serviceaccount`

`istio-csr` CR object will be a singleton object with the name `default` and an OpenShift user will have to
create the same for enabling the deployment. `istio-csr` should be deployed in the same namespace as that of
the `OpenShift Service Mesh` control plane.

> When `istiocsr` custom resource is created but `issuerref` is not provided operator will create a

Reviewer: Can you take a relook at this paragraph? cert-manager Certficiate either uses a cluster scoped issuer (ClusterIssuer) or namespace scoped issuer (referred as only Issuer). There is missing context about the Issuer being created using ClusterIssuer.

Author: Yeah it was confusing. Rephrased the content.

`ClusterIssuer` in the `cert-manager` namespace since cert-manager controller observes CAs created in its own
namespace. And using the ClusterIssuer an Issuer will be created in the istio-csr's namespace for signing the
certificates requested by `OpenShift Service Mesh`.

An OpenShift user could update any configurations in the spec section of the API, except for below
configurations which are immutable and once set cannot be modified.
- namespace:

>   Namespace where the `istio-csr` agent will be deployed, will also be used for creating other resources

Author: We are making fields immutable by having CEL validations to check if the content has been updated.

  required for agent to be functional and the same namespace will be used for creating the `issuer` which
  is used for signing the certificate requests. And allowing to change this config will disrupt the
  functionality of `OpenShift Service Mesh` also deployed in the same namespace.
- issuerRef:
  IssuerRef is the reference to the certificate authority used for signing the certificates requested by

>   `OpenShift Service Mesh`. And allowing to change this config will create issues to renewing the existing

Author: This content has been removed, since changing issuer in istio-csr or istiod is a valid scenario.

  certificates.

Configurations made available in the spec of `istio-csr` custom resource are passed as process arguments and
updating these configurations would cause new rollout to happen.

> When an OpenShift user deletes istio-csr CR object `istio-csr` agent deployment will be deleted and all the

Author: Controller will retry to delete the resources created by it.

> When an OpenShift user deletes istio-csr CR object `istio-csr` agent deployment will be deleted and all the

Reviewer: What is the expected behaviour when istio-csr deployment is deleted? I think, there might be service disruption when istio tries to get new/renew certificate. istio-csr operand should not be allowed to delete if there is a reference being set in istio.

> When an OpenShift user deletes istio-csr CR object `istio-csr` agent deployment will be deleted and all the

Author: Yes, it will cause a service disruption. But would it be correct for istio-csr controller to have that dependency on istio? I think it should not be.

other resources created will also be cleaned up.

A fork of [upstream](https://github.com/cert-manager/istio-csr) `istio-csr` will be created [downstream](https://github.com/openshift)
for better version management.

### Workflow Description

> - An OpenShift user enables `istio-csr` by creating the new custom resource made available for configuring

Author: I think the sequence diagram will just depict the normal controller flow of user provisioning the controller CR, which will make controller to deploy the operand. Please let me know if its really needed and what specifics should be included in the diagram.

> - An OpenShift user enables `istio-csr` by creating the new custom resource made available for configuring

Reviewer: The sequence diagram will help in having more clarity on the e2e flow from user installation to initial certifcate setup. Also, the delete usecases.

> - An OpenShift user enables `istio-csr` by creating the new custom resource made available for configuring

Author: suggestion incorporated.

	Port string `json:"port,omitempty"`

	// CertificateKeySize is the server's serving certificate's key size.
	// +kubebuilder:default:=2048
	// +optional
	CertificateKeySize int32 `json:"certificateKeySize,omitempty"`

	// SignatureAlgorithm is the signature algorithm to use when generating
	// private keys. At present only RSA and ECDSA are supported.
	// +kubebuilder:default:="RSA"
	// +kubebuilder:validation:Enum:="RSA";"ECDSA"
	// +optional
	SignatureAlgorithm string `json:"signatureAlgorithm,omitempty"`

	// MaxCertificateDuration is the maximum validity duration that can be
	// requested for a certificate.
	// +kubebuilder:default:="1h"
	// +optional
	MaxCertificateDuration time.Duration `json:"certificateDuration,omitempty"`
}

// IstioConfig is for configuring the istio specifics.
type IstioConfig struct {
	// Revisions are the istio revisions that are currently installed in the cluster.
	// Changing this field will modify the DNS names that will be requested for
	// the istiod certificate.
	// +kubebuilder:default:=["default"]
	// +optional
	Revisions []string `json:"revisions,omitempty"`
}

// IstioCSRStatus is the most recently observed status of the IstioCSR.

> type IstioCSRStatus struct {

Reviewer: A re-look at status fields of operator and controller will be needed as per new context.

Author: @TrilokGeer I think offline discussion we had to aggregate the state of all operands managed by the operator will be a different proposal, and here the `IstioCSRStatus` will just hold the details of deployed `istio-csr` operand. Please let me know WDYT?

  with default istio-csr configuration.
- Enable istio-csr by creating the istio-csr CR object with permutations of
  configurations and validate the behavior.
- Upgrade/downgrade testing
  - Scenarios mentioned in the section Upgrade / Downgrade Strategy has expected behavior.
  - Sufficient time for feedback from the QE.
- The feature is available by default and does not have any specific featureGate defined.

## Graduation Criteria

### Dev Preview -> Tech Preview

- Feature available for end-to-end usage.
- Complete end user documentation.
- UTs and e2e tests are present.
- Gather feedback from the users.

### Tech Preview -> GA

N/A. This feature is for Dev Preview, until decided for GA.

### Removing a deprecated feature

None.

## Upgrade / Downgrade Strategy

On upgrade:
- cert-manager-operator will have functionality to enable istio-csr agent and based
  on the administrator configuration, agent will be deployed and available for usage
  with Service Mesh.
- Enabling istio-csr when Service Mesh was already deployed before upgrade is not
  [supported](https://cert-manager.io/docs/usage/istio-csr/#installing-istio-csr-after-istio).

## Version Skew Strategy

> Istio-csr will be supported for OpenShift Service Mesh Operator 2.4+, Istio v1.10+ and cert-manager v1.3+.

Author: comment incorporated.

	// More info: https://git.k8s.io/community/contributors/devel/sig-architecture/api-conventions.md#metadata
	metav1.ObjectMeta `json:"metadata,omitempty"`

	// spec is the specification of the desired behavior of the IstioCSR.
	Spec IstioCSRSpec `json:"spec,omitempty"`

	// status is the most recently observed status of the IstioCSR.
	Status IstioCSRStatus `json:"status,omitempty"`
}

// IstioCSRSpec is the specification of the desired behavior of the IstioCSR.
type IstioCSRSpec struct {
	apiv1.OperatorSpec `json:",inline"`

	// ControllerConfig is for configuring istio-csr operator behavior.
	ControllerConfig `json:"controllerConfig,omitempty"`

	// ApplicationConfig is for configuring istio-csr operand behavior.
	ApplicationConfig `json:"applicationConfig,omitempty"`
}

// ControllerConfig is for configuring istio-csr operator behavior.
type ControllerConfig struct {
	// Namespace where the istio control-plane is running.
	// Same namespace will be used for deploying istio-csr agent.
	// +kubebuilder:default:="istio-system"
	// +kubebuilder:validation:XValidation:rule="oldSelf == '' || self == oldSelf",message="namespace is immutable once set"
	// +optional

> 	Namespace string `json:"namespace,omitempty"`

Author: But `Namespace` is for installing the operand `istio-csr`, the controller is managing. And thought not assume the `istiod` namespace, made it as a config param. Please let me know WDYT.

> 	Namespace string `json:"namespace,omitempty"`

Reviewer 2: I am not sure that I understand it correctly. If I create `IstioCSR` resource, will it result in deploying a controller that will manage `istio-csr` agent? And the controller will be created in `IstioCSR.metadata.namespace` and the agent in `IstioCSR.spec.controllerConfig.namespace`?

> 	Namespace string `json:"namespace,omitempty"`

Author: Yeah the field name `ControllerConfig` is misleading, I will update it.

> the controller will be created in IstioCSR.metadata.namespace and the agent in IstioCSR.spec.controllerConfig.namespace

istio-csr manager is part of `cert-manager-operator` which will watch for `istiocsr` custom resource and install `istio-csr` agent in `IstioCSR.spec.controllerConfig.namespace`. In fact all the configs in `controllerConfig` is for setting up the agent. I will update the name and comments appropriately.
And I think it would be better to add a CEL validation for restricting the `istiocsr` custom resource in cert-manager-operator namespace. Please let me know WDYT.

> 	Namespace string `json:"namespace,omitempty"`

Reviewer 2: If istio-csr agent must run in the same namespace as istiod, I would remove `controllerConfig.namespace` and rely only on the `metadata.namespace`.

> 	Namespace string `json:"namespace,omitempty"`

Author: Yeah, but the expectation is to create the new CR in operator namespace and namespace where istiod will be installed need not exist at the time, operator will create it.

> 	Namespace string `json:"namespace,omitempty"`

Reviewer 2: > namespace where istiod will be installed need not exist at the time

Why?

> 	Namespace string `json:"namespace,omitempty"`

Author: My understanding is based on this [doc](https://cert-manager.io/docs/usage/istio-csr/#installing-istio-csr-after-istio) that, istio-csr can be and must be installed before istio. Please correct me if understanding is wrong.

	// +kubebuilder:default:="5s"
	// +optional
	ScrapeTimeout time.Duration `json:"scrapeTimeout,omitempty"`

	// Labels to add to the ServiceMonitor resource.
	// +optional
	Labels map[string]string `json:"labels,omitempty"`
}

// CertManagerConfig is for configuring cert-manager specifics.
type CertManagerConfig struct {
	// IssuerRef contains details to the referenced object used for
	// obtaining the certificates.
	// When unset operator will create Issuer in the configured istio's
	// namespace.
	// +kubebuilder:validation:XValidation:rule="oldSelf == '' || self == oldSelf",message="issuerRef is immutable once set"
	// +optional
	IssuerRef certmanagerv1.ObjectReference `json:"issuerRef,omitempty"`
}

// TLSConfig is for configuring certificate specifics.
type TLSConfig struct {
	// TrustDomain is the cluster's trust domain.
	// +kubebuilder:default:="cluster.local"
	// +optional
	TrustDomain string `json:"trustDomain,omitempty"`

	// RootCAFile is for setting the file location containing the root CA which is
	// present in the configured IssuerRef. File should be made available using the
	// Volume and VolumeMount options.

> 	// +kubebuilder:default:="/var/run/secrets/istio-csr/ca.crt"

Reviewer 2: We shouldn't set the default value for this field if we do not require mounting static root CA.
```suggestion
```

Author: Will remove the default.

Author: suggestion restored.

	// +kubebuilder:default:="0.0.0.0"
	// +optional
	Address string `json:"address,omitempty"`

	// Port to serve istio-csr gRPC service.
	// +kubebuilder:default:="6443"
	// +optional
	Port string `json:"port,omitempty"`

	// CertificateKeySize is the server's serving certificate's key size.
	// +kubebuilder:default:=2048
	// +optional
	CertificateKeySize int32 `json:"certificateKeySize,omitempty"`

	// SignatureAlgorithm is the signature algorithm to use when generating
	// private keys. At present only RSA and ECDSA are supported.
	// +kubebuilder:default:="RSA"
	// +kubebuilder:validation:Enum:="RSA";"ECDSA"
	// +optional
	SignatureAlgorithm string `json:"signatureAlgorithm,omitempty"`

	// MaxCertificateDuration is the maximum validity duration that can be
	// requested for a certificate.
	// +kubebuilder:default:="1h"
	// +optional
	MaxCertificateDuration time.Duration `json:"certificateDuration,omitempty"`
}

// IstioConfig is for configuring the istio specifics.
type IstioConfig struct {
	// Revisions are the istio revisions that are currently installed in the cluster.
	// Changing this field will modify the DNS names that will be requested for
	// the istiod certificate.

> 	// +kubebuilder:default:=["basic"]

Reviewer: The change to default does not refect in the PR diff. Is it intentional?

> 	// +kubebuilder:default:=["basic"]

Author: No, maybe its issue from my side during rebase.

      apiVersion: monitoring.coreos.com/v1
      kind: ServiceMonitor
      metadata:
        name: cert-manager-istio-csr
        namespace: istio-system
        labels:
          app: cert-manager-istio-csr
          prometheus: default
      spec:
        jobLabel: cert-manager-istio-csr
        selector:
          matchLabels:
            app: cert-manager-istio-csr-metrics
        namespaceSelector:
          matchNames:
            - cert-manager
        endpoints:
        - targetPort: 9402
          path: "/metrics"
          interval: 10s
          scrapeTimeout: 5s
      ```

   5. ServiceAccount required by the agent, which will be associated with the required ClusterRole and
      Roles in next step.
      ```yaml
      apiVersion: v1
      kind: ServiceAccount
      metadata:
        labels:
          app: cert-manager-istio-csr
        name: cert-manager-istio-csr
        namespace: istio-system
      ```

   6. ClusterRoles and Roles required by istio-csr.
      ```yaml
      ## required for creating ConfigMap with the CA in selected namespaces
      ## by istio-csr agent.
      kind: ClusterRole
      apiVersion: rbac.authorization.k8s.io/v1
      metadata:

>         name: cert-manager-istio-csr

Reviewer 2: We may need a suffix for cluster role names to support multi-tenancy in the future.

  when required by creating the custom resource.
- As an OpenShift user, I want to have an option to dynamically configure `istio-csr`, so that only the required
  features can be enabled by updating the custom resource.
- As an OpenShift user, I should be able to disable `istio-csr` when not required by removing the custom resource,
  and controller should cleanup all resources created for the `istio-csr` deployment.
- As an OpenShift user, I should be able to install `istio-csr` in the upgrade cluster where istiod control plane
  is active and should be able to update the certificate endpoint to `istio-csr` endpoint.
- As an OpenShift user, I want to have an option to dynamically enable monitoring for the `istio-csr` project and
  to use the OpenShift monitoring solution when required.

### Goals

- `cert-manager-operator` to be extended to manage `istio-csr` along with currently managed `cert-manager`.
- New custom resource `istiocsr` to be made available to enable, configure and disable deployment.

### Non-Goals

- `istio-csr` can be used only with supported version of `OpenShift Service Mesh`. Please refer `Version Skew Strategy`
  section for more details.

## Proposal

> `istio-csr` will be managed by `cert-manager-operator` which can be enabled and configured by a user as and

Author: Image version and tag will configured as a environment variable in operator [deployment](https://github.com/openshift/cert-manager-operator/blob/master/config/manager/manager.yaml#L60-L67) spec. Currently downstream repo is synced to [v0.12.0](https://github.com/openshift/cert-manager-istio-csr/releases/tag/v0.12.0) and using the same downstream image will be created.

type IstioCSR struct {
	metav1.TypeMeta `json:",inline"`

	// metadata is the standard object's metadata.
	// More info: https://git.k8s.io/community/contributors/devel/sig-architecture/api-conventions.md#metadata
	metav1.ObjectMeta `json:"metadata,omitempty"`

	// spec is the specification of the desired behavior of the IstioCSR.
	Spec IstioCSRSpec `json:"spec,omitempty"`

	// status is the most recently observed status of the IstioCSR.
	Status IstioCSRStatus `json:"status,omitempty"`
}

// IstioCSRSpec is the specification of the desired behavior of the IstioCSR.
type IstioCSRSpec struct {
	// istioCSRConfig is for configuring the istio-csr agent behavior.
	IstioCSRConfig *IstioCSRConfig `json:"istioCSRConfig,omitempty"`

	// controllerConfig is for configuring the controller for setting up
	// defaults to enable istio-csr agent.
	ControllerConfig *ControllerConfig `json:"controllerConfig,omitempty"`
}

// IstioCSRConfig is for configuring the istio-csr agent behavior.
type IstioCSRConfig struct {
	// replicas is the number of desired istio-csr agent replicas.
	// +kubebuilder:default:=1
	// +optional

> 	Replicas int32 `json:"replicas,omitempty"`

Author: Will remove the field.

> 	Replicas int32 `json:"replicas,omitempty"`

Reviewer: Is there a plan to implement a webhook?

> 	Replicas int32 `json:"replicas,omitempty"`

Author: istio-csr has a webhook implementation.

	// logLevel is for setting verbosity of istio-csr agent logging.
	// Supported log levels: 1-5.
	// +kubebuilder:default:=1
	// +kubebuilder:validation:Minimum:=1
	// +kubebuilder:validation:Maximum:=5
	// +optional
	LogLevel int32 `json:"logLevel,omitempty"`

	// logFormat is for specifying the output format of istio-csr agent logging.
	// Support log formats are text and json.
	// +kubebuilder:default:=text
	// +optional
	LogFormat string `json:"logFormat,omitempty"`

	// metrics is for configuring Prometheus metrics behavior.
	// +optional

> 	Metrics *MetricsConfig `json:"metrics,omitempty"`

Author: Will remove the config field.

> 	Metrics *MetricsConfig `json:"metrics,omitempty"`

Author: suggestion incorporated.

	// certmanager is for configuring cert-manager specifics.
	// +optional

> 	CertManager *CertManagerConfig `json:"certmanager,omitempty"`

Author: Will make it mandatory.

> 	CertManager *CertManagerConfig `json:"certmanager,omitempty"`

Author: suggestion incorporated.

	// ref: https://kubernetes.io/docs/concepts/scheduling-eviction/taint-and-toleration/
	// +optional
	Tolerations []corev1.Toleration `json:"tolerations,omitempty"`

	// NodeSelector is for defining the scheduling criteria using node labels.
	// ref: https://kubernetes.io/docs/concepts/configuration/assign-pod-node/
	// +optional
	NodeSelector map[string]string `json:"nodeSelector,omitempty"`
}

// MetricsConfig is for configuring Prometheus metrics behavior.
type MetricsConfig struct {
	// port for exposing Prometheus metrics on 0.0.0.0 on path '/metrics'.
	// +kubebuilder:default:="9402"
	// +optional
	Port string `json:"port,omitempty"`

	// service is for configuring the metrics server specifics.
	// +optional
	Service *MetricServiceConfig `json:"service,omitempty"`
}

// MetricServiceConfig is for confuring the metrics server specifics.
type MetricServiceConfig struct {
	// enabled when set will create a Service resource to expose metrics endpoint.
	// +kubebuilder:default:=false
	// +optional
	Enabled bool `json:"enabled,omitempty"`

	// serviceType of the service resource created for metrics.
	// +kubebuilder:default:="ClusterIP"
	// +optional

> 	ServiceType corev1.ServiceType `json:"serviceType,omitempty"`

Author: Will remove the field.

> 	ServiceType corev1.ServiceType `json:"serviceType,omitempty"`

Author: suggestion incorporated.

	// serviceMonitor is for configuring the service monitor specifics
	// for the metrics service.
	// +optional
	ServiceMonitor *MetricsServiceMonitorConfig `json:"serviceMonitor,omitempty"`
}

// MetricsServiceMonitorConfig is for configuring the service monitor specifics
// for the metrics service.
type MetricsServiceMonitorConfig struct {
	// enabled when set creates Prometheus ServiceMonitor resource for approver-policy.
	// +kubebuilder:default:=false
	// +optional
	Enabled bool `json:"enabled,omitempty"`

	// prometheusInstance name to set the value for the "prometheus" label on the ServiceMonitor.
	// +kubebuilder:default:="default"
	// +optional

> 	PrometheusInstance string `json:"prometheusInstance,omitempty"`

Author: This config will be removed based on [suggestion](https://github.com/openshift/enhancements/pull/1628#discussion_r1832182236)

> 	PrometheusInstance string `json:"prometheusInstance,omitempty"`

Author: suggestion incorporated.

	// interval defines periodicity the Prometheus will scrape for metrics.
	// +kubebuilder:default:="10s"
	// +optional
	Interval time.Duration `json:"interval,omitempty"`

	// scrapeTimeout defines timeout on each metric probe request.
	// +kubebuilder:default:="5s"
	// +optional
	ScrapeTimeout time.Duration `json:"scrapeTimeout,omitempty"`

	// labels to add to the ServiceMonitor resource.
	// +optional
	Labels map[string]string `json:"labels,omitempty"`
}

// CertManagerConfig is for configuring cert-manager specifics.
type CertManagerConfig struct {
	// issuerRef contains details to the referenced object used for
	// obtaining the certificates.

> 	// When unset operator will create Issuer in the configured istio's

Author: It's provided as an option. If its best to mandate user to configure the issue, will update config.

> 	// When unset operator will create Issuer in the configured istio's

Author: suggestion incorporated.

	// namespace.
	// +kubebuilder:validation:XValidation:rule="oldSelf == '' || self == oldSelf",message="issuerRef is immutable once set"
	// +optional

> 	IssuerRef certmanagerv1.ObjectReference `json:"issuerRef,omitempty"`

Author: Will make it mandatory.

> 	IssuerRef certmanagerv1.ObjectReference `json:"issuerRef,omitempty"`

Author: suggestion incorporated.

}

// TLSConfig is for configuring certificate specifics.

> type TLSConfig struct {

Author: Will rename the config.

> type TLSConfig struct {

Author: Will mandate the config for user.

> type TLSConfig struct {

Author: suggestion incorporated.

> type TLSConfig struct {

Author: suggestion incorporated.

	// +kubebuilder:default:=2048
	// +optional
	CertificateKeySize int32 `json:"certificateKeySize,omitempty"`

	// SignatureAlgorithm is the signature algorithm to use when generating
	// private keys. At present only RSA and ECDSA are supported.
	// +kubebuilder:default:="RSA"
	// +kubebuilder:validation:Enum:="RSA";"ECDSA"
	// +optional
	SignatureAlgorithm string `json:"signatureAlgorithm,omitempty"`

	// MaxCertificateDuration is the maximum validity duration that can be
	// requested for a certificate.
	// +kubebuilder:default:="1h"
	// +optional
	MaxCertificateDuration time.Duration `json:"certificateDuration,omitempty"`
}

// IstioConfig is for configuring the istio specifics.
type IstioConfig struct {
	// Revisions are the istio revisions that are currently installed in the cluster.
	// Changing this field will modify the DNS names that will be requested for
	// the istiod certificate.
	// +kubebuilder:default:=["basic"]
	// +optional
	Revisions []string `json:"revisions,omitempty"`

	// namespace of the istio control-plane. In the same namespace issuer will be created
	// used for obtaining the serving certificates.

> 	Namespace string `json:"namespace,omitempty"`

Reviewer: istio control-plane namespace should be mandatory.

Author: Will make it mandatory.

Reviewer: What is the rbac configuration needed to install in the istio control plane namespace?

Author: Just for the reference, it was initially planned as markers, this will be added as manifest templates to create on demand.
```yaml
//+kubebuilder:rbac:groups="cert-manager.io",resources=clusterissuer,verbs=get;list;watch;create;delete
//+kubebuilder:rbac:groups="cert-manager.io",namespace=istio-system,resources=certificate;issuer,verbs=get;list;watch;create;delete
//+kubebuilder:rbac:groups="",resources=events;services;namespaces;serviceaccounts;configmaps,verbs=get;list;watch;create;update;patch;delete
//+kubebuilder:rbac:groups="rbac.authorization.k8s.io",resources=clusterroles;clusterrolebindings,verbs=get;list;watch;create;update;patch;delete
//+kubebuilder:rbac:groups="rbac.authorization.k8s.io",namespace=istio-system,resources=roles;rolebindings,verbs=get;list;watch;create;update;patch;delete
```

}

// ControllerConfig is for configuring the controller for setting up
// defaults to enable istio-csr agent.
type ControllerConfig struct {
	// labels to apply to all resources created for istio-csr agent deployment.

> 	Labels map[string]string `json:"labels,omitempty"`

Reviewer: A default set of labels would be required for resources created by istio-csr for ease of query and server-side filtering.

Author: We are adding a default set of labels, this field is for adding user specific labels.

Reviewer: Can you mention explicitly in the proposal of the default set of labels and their significance ?

	// issuerConfig is for setting the issuer specifics to be created by the controller.

> 	IssuerConfig *IssuerConfig `json:"issuerConfig,omitempty"`

Reviewer: Is this issuer created automatically by istio-csr as default or operator? It'd be best to let user add the CR rather than replicating the same here, if possible.

Author: suggestion incorporated.

}

type IssuerConfig struct {
	// issuerType is for specifying issuer type `ClusterIssuer` or `Issuer`.
	Issuertype string `json:"issuerType,omitempty"`

	// duration is for setting the default certificate validity duration.
	Duration string `json:"duration,omitempty"`

	// renewBefore is for specifying the certificate renewal time before the certificate expiry.
	RenewBefore string `json:"renewBefore,omitempty"`
}

// IstioCSRStatus is the most recently observed status of the IstioCSR.

> type IstioCSRStatus struct {

Reviewer: Istio csr status should reflect the present image version , tag deployed and deployment details.

Author: Will add the info to status.

Author: suggestion incorporated.

	// conditions holds information of the current state of the istio-csr agent deployment.

> 	Conditions *metav1.Condition `json:"conditions,omitempty"`

Reviewer: Any status that indicates failure, should set the operator status to degraded disabling upgrades or any further add-on operations.

Author: Conditions will indicate degrade status if there is any unrecoverable error.

Author: Yes, if any of the required resources are not in expected state, `IstioCSRStatus` condition will be updated with error observed in the problematic resource.

Reviewer: Can you explicitly add the statements in the proposal?

Reviewer: Also, the scenario where istio-csr CR is removed, the expected operator status should be updated.

	// generatedIstioCSRConfig holds information about the clusterissuer/issuer created
	// by the controller or user required by the istio-csr agent.

> 	GeneratedIstioCSRConfig []IstioCSRConfig `json:"generatedIstioCSRConfig,omitempty"`

Reviewer: Ideally, all cert-manager specific resources related to issuers must be created by user.

Author: This will be used when operator creates the issuer. Based on the suggestions since we are enforcing user to configure issue, this will be updated.

Author: suggestion incorporated.

- `.spec.istioCSRConfig.istio.namespace`

When an OpenShift user deletes `istiocsr.operator.openshift.io` CR object, `istio-csr-controller` will stop managing all
the resources created for installing `istio-csr` agent and user will manually clean up the resources. Please refer
`Operational Aspects of API Extensions` section for command to list all the resources.

`istiocsr.operator.openshift.io` CR status sub-resource will be used for updating the status of the `istio-csr` agent
installation, any error encountered while creating the required resources or the reconciling the state.

A fork of [upstream](https://github.com/cert-manager/istio-csr) `istio-csr` will be created [downstream](https://github.com/openshift/cert-manager-istio-csr)
for better version management.

### Workflow Description

- Enable `istio-csr-controller`
  - An OpenShift user enables `istio-csr-controller` by creating the `istiocsr.operator.openshift.io` CR.
  - `istio-csr-controller` based on the configuration in `istiocsr.operator.openshift.io` CR, deploys `istio-csr` agent
    in the CR created namespace.

![alt text](./istio-csr-create.png).

- Disable `istio-csr-controller`

>   - An OpenShift user disables `istio-csr-controller` by deleting the `istiocsr.operator.openshift.io` CR.

Reviewer: User is not aware of the controller in the operator. Maybe this should read as , istio-csr-controller is stopped on deletion of the CR.
